src/database/size.ts

```diff
@@ -0,0 +1,115 @@
+import * as logger from "../logger";
+import * as pathLib from "path";
+
+import { ListRemote, RTDBListRemote } from "./listRemote";
+import { RTDBSizeRemote, SizeRemote } from "./sizeRemote";
+import { SizeResult } from "./sizeResult";
+import { Stack } from "../throttler/stack";
+
+/*
+ * For flat objects, performance suffers due to "slow-start"
+ * shallow gets. We set a generous initial list batch size
+ * to mitigate this.
+ */
+const LIST_BATCH_SIZE = 4000;
+
+/*
+ * Top-level GET requests for large subtrees will likely
+ * fail, so there are some probable gains to be had from
+ * performing shallow gets first. This constant controls
+ * how deep to search in the tree before attempting large
+ * GETs.
+ */
+const SKIP_DEPTH = 1;
+const TIMEOUT = 50;
+const TIMEOUT_STATUS_CODE = 400;
+
+/*
+ * Control how many outstanding full GETs and shallow
+ * GETs are allowed at a time. Wide trees will have
+ * substantially more "size" operations, whereas deep
+ * trees will have substantially more outstanding list
+ * operations.
+ */
+const SIZE_STACK_CONCURRENCY = 2000;
+const LIST_STACK_CONCURRENCY = 2000;
+
+export class DatabaseSize {
+  path: string;
+  skipDepth: number;
+  sizeEstimate: number;
+
+  listRemote: ListRemote;
+  sizeRemote: SizeRemote;
+
+  private listStack: Stack<() => Promise<string[]>, string[]>;
+  private sizeStack: Stack<() => Promise<SizeResult>, SizeResult>;
+
+  constructor(instance: string, path: string, depth?: number) {
+    this.path = path;
+    this.skipDepth = depth || SKIP_DEPTH;
+    this.sizeEstimate = 0;
+    this.sizeRemote = new RTDBSizeRemote(instance);
+    this.listRemote = new RTDBListRemote(instance);
+    this.listStack = new Stack({
+      name: "list stack",
+      concurrency: LIST_STACK_CONCURRENCY,
+      retries: 3,
+    });
+    this.sizeStack = new Stack({
+      name: "size stack",
+      concurrency: SIZE_STACK_CONCURRENCY,
+      retries: 3,
+    });
+  }
+
+  async execute(): Promise<number> {
+    return this.getSubtreeSize(this.path, 0);
+  }
+
+  private async getSubtreeSize(path: string, depth: number): Promise<number> {
+    let sizeEstimate = 0;
+
+    /*
+     * If we are sufficiently deep in the JSON subtree, try sizing the
+     * entire subtree in a single GET.
+     */
+    if (depth >= this.skipDepth) {
+      const result: SizeResult = await this.sizeStack.run(() =>
+        this.sizeRemote.sizeNode(path, TIMEOUT)
+      );
+      if (result !== null && result.success) {
+        sizeEstimate = result.bytes;
+        return sizeEstimate;
+      }
+      /*
+       * Ignore timeout errors so we can try sizing children of this node.
+       */
+      if (!result.success && result.err.status !== TIMEOUT_STATUS_CODE) {
+        logger.debug(
+          `Unexpected error: '${result.err.message}' when sizing node ${path}. Ignoring.`
+        );
+        throw result.err;
+      }
+    }
+
+    let subPaths: string[] = [];
+    let offset: string;
```

Comment:
this is a bit non-standard for typescript, is `null` a valid offset or just a placeholder?

Reply:
In this case it will just be a placeholder. The `listPath` method in `src/database/listRemote.ts` accepts it as an optional argument. The default value that an optional argument which a caller doesn't specify takes is "undefined". Also, according to the documentation default values for optional parameters will only be assumed if the value of the argument is undefined (be it by caller omission, or the caller explicitly passing undefined) [1]. Therefore, I think it makes sense to leave this undefined. I will add the comment though.

[1] https://www.typescriptlang.org/docs/handbook/functions.html#optional-and-default-parameters